hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequest.java

```diff
@@ -0,0 +1,348 @@
+import org.apache.hadoop.ozone.om.request.key.OMKeyRequest;
+import org.apache.hadoop.ozone.om.response.OMClientResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .CreateFileRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .KeyArgs;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMRequest;
+import org.apache.hadoop.ozone.security.acl.IAccessAuthorizer;
+import org.apache.hadoop.ozone.security.acl.OzoneObj;
+import org.apache.hadoop.util.Time;
+import org.apache.hadoop.utils.UniqueId;
+import org.apache.hadoop.utils.db.Table;
+import org.apache.hadoop.utils.db.TableIterator;
+import org.apache.hadoop.utils.db.cache.CacheKey;
+import org.apache.hadoop.utils.db.cache.CacheValue;
+
+
+import static org.apache.hadoop.ozone.om.request.file.OMFileRequest.OMDirectoryResult.DIRECTORY_EXISTS;
+import static org.apache.hadoop.ozone.om.request.file.OMFileRequest.OMDirectoryResult.DIRECTORY_EXISTS_IN_GIVENPATH;
+import static org.apache.hadoop.ozone.om.request.file.OMFileRequest.OMDirectoryResult.FILE_EXISTS_IN_GIVENPATH;
+import static org.apache.hadoop.ozone.om.request.file.OMFileRequest.OMDirectoryResult.FILE_EXISTS;
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.BUCKET_LOCK;
+import static org.apache.hadoop.ozone.om.request.file.OMFileRequest.OMDirectoryResult.NONE;
+
+/**
+ * Handles create file request.
+ */
+public class OMFileCreateRequest extends OMKeyCreateRequest
+    implements OMKeyRequest {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMFileCreateRequest.class);
+  public OMFileCreateRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    CreateFileRequest createFileRequest = getOmRequest().getCreateFileRequest();
+    Preconditions.checkNotNull(createFileRequest);
+
+    KeyArgs keyArgs = createFileRequest.getKeyArgs();
+
+    if (keyArgs.getKeyName().length() == 0) {
+      // Check if this is the root of the filesystem.
+      // Not throwing exception here, as need to throw exception after
+      // checking volume/bucket exists.
+      return getOmRequest().toBuilder().setUserInfo(getUserInfo()).build();
+    }
+
+    long scmBlockSize = ozoneManager.getScmBlockSize();
+
+    // NOTE size of a key is not a hard limit on anything, it is a value that
+    // client should expect, in terms of current size of key. If client sets
+    // a value, then this value is used, otherwise, we allocate a single
+    // block which is the current size, if read by the client.
+    final long requestedSize = keyArgs.getDataSize() > 0 ?
+        keyArgs.getDataSize() : scmBlockSize;
+
+    boolean useRatis = ozoneManager.shouldUseRatis();
+
+    HddsProtos.ReplicationFactor factor = keyArgs.getFactor();
+    if (factor == null) {
+      factor = useRatis ? HddsProtos.ReplicationFactor.THREE :
```

Comment:
Can we have DataNode with single-node Ratis?

Reply:
Yes. With Ratis, replication factor 1 and 3 are supported.

Reply:
Then is this logic correct? We cannot use Ratis to infer that replication factor is 3.

Reply:
As the default value for useRatis is false, this logic will work. But i think we should revisit this logic.
as previously for replication factor 1 we used to use STAND_ALONE, but RATIS can be used with 1 and 3.

Reply:
I think we can eliminate this property completely if these settings are not passed, we can use RATIS and ONE.

Reply:
Reached out to @arp7 as existing Ozone NON-HA code is also behaving like this. We will proceed with, for now, will open a new Jira to address this.